testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java

```diff
@@ -0,0 +1,927 @@
+        .messageRead(any(InputStream.class));
+    verify(mockClientStreamListener, timeout(TIMEOUT_MS).times(clientSent + 5))
+        .messageRead(any(InputStream.class));
+
+    // And now check that the streams can still complete gracefully
+    clientStream.writeMessage(methodDescriptor.streamRequest(largeMessage));
+    clientStream.flush();
+    clientStream.halfClose();
+    doPingPong(serverListener);
+    verify(mockServerStreamListener, never()).halfClosed();
+
+    serverStream.request(1);
+    verify(mockServerStreamListener, timeout(TIMEOUT_MS).times(serverSent + 6))
+        .messageRead(any(InputStream.class));
+    verify(mockServerStreamListener, timeout(TIMEOUT_MS)).halfClosed();
+
+    serverStream.writeMessage(methodDescriptor.streamResponse(largeMessage));
+    serverStream.flush();
+    Status status = Status.OK.withDescription("... quite a lengthy discussion");
+    serverStream.close(status, new Metadata());
+    doPingPong(serverListener);
+    verify(mockClientStreamListener, never()).closed(any(Status.class), any(Metadata.class));
+
+    clientStream.request(1);
+    verify(mockClientStreamListener, timeout(TIMEOUT_MS).times(clientSent + 6))
+        .messageRead(any(InputStream.class));
+    verify(mockServerStreamListener, timeout(TIMEOUT_MS)).closed(statusCaptor.capture());
+    assertCodeEquals(Status.OK, statusCaptor.getValue());
+    verify(mockClientStreamListener, timeout(TIMEOUT_MS))
+        .closed(statusCaptor.capture(), any(Metadata.class));
+    assertEquals(status.getCode(), statusCaptor.getValue().getCode());
+    assertEquals(status.getDescription(), statusCaptor.getValue().getDescription());
+  }
+
+  /**
+   * Helper that simply does an RPC. It can be used similar to a sleep for negative testing: to give
+   * time for actions _not_ to happen. Since it is based on doing an actual RPC with actual
+   * callbacks, it generally provides plenty of time for Runnables to execute. But it is also faster
+   * on faster machines and more reliable on slower machines.
+   */
+  private void doPingPong(MockServerListener serverListener) throws InterruptedException {
+    ManagedClientTransport client = newClientTransport();
+    client.start(mock(ManagedClientTransport.Listener.class));
+    ClientStream clientStream = client.newStream(methodDescriptor, new Metadata());
+    ClientStreamListener mockClientStreamListener = mock(ClientStreamListener.class);
+    clientStream.start(mockClientStreamListener);
+
+    MockServerTransportListener serverTransportListener
+        = serverListener.takeListenerOrFail(TIMEOUT_MS, TimeUnit.MILLISECONDS);
+    StreamCreation serverStreamCreation
+        = serverTransportListener.takeStreamOrFail(TIMEOUT_MS, TimeUnit.MILLISECONDS);
+    ServerStream serverStream = serverStreamCreation.stream;
+    ServerStreamListener mockServerStreamListener = serverStreamCreation.listener;
+
+    serverStream.close(Status.OK, new Metadata());
```

Comment:
So we're just ping-ponging headers? Seems fine, just not what I expected for `pingpong`.